src/flow/src/batching_mode/engine.rs

```diff
@@ -0,0 +1,306 @@
+pub struct BatchingEngine {
+    tasks: RwLock<BTreeMap<FlowId, BatchingTask>>,
+    shutdown_txs: RwLock<BTreeMap<FlowId, oneshot::Sender<()>>>,
+    frontend_client: Arc<FrontendClient>,
+    flow_metadata_manager: FlowMetadataManagerRef,
+    table_meta: TableMetadataManagerRef,
+    query_engine: QueryEngineRef,
+}
+
+impl BatchingEngine {
+    pub fn new(
+        frontend_client: Arc<FrontendClient>,
+        query_engine: QueryEngineRef,
+        flow_metadata_manager: FlowMetadataManagerRef,
+        table_meta: TableMetadataManagerRef,
+    ) -> Self {
+        Self {
+            tasks: Default::default(),
+            shutdown_txs: Default::default(),
+            frontend_client,
+            flow_metadata_manager,
+            table_meta,
+            query_engine,
+        }
+    }
+
+    pub async fn handle_inserts(
+        &self,
+        request: api::v1::region::InsertRequests,
+    ) -> Result<FlowResponse, Error> {
+        let table_info_mgr = self.table_meta.table_info_manager();
+        let mut group_by_table_id: HashMap<TableId, Vec<api::v1::Rows>> = HashMap::new();
+
+        for r in request.requests {
+            let tid = RegionId::from(r.region_id).table_id();
+            let entry = group_by_table_id.entry(tid).or_default();
+            if let Some(rows) = r.rows {
+                entry.push(rows);
+            }
+        }
+
+        let tids = group_by_table_id.keys().cloned().collect::<Vec<TableId>>();
+        let table_infos =
+            table_info_mgr
+                .batch_get(&tids)
+                .await
+                .with_context(|_| TableNotFoundMetaSnafu {
+                    msg: format!("Failed to get table info for table ids: {:?}", tids),
+                })?;
+
+        if !tids.iter().all(|id| table_infos.contains_key(id)) {
+            warn!(
+                "Failed to get all the table info for table ids, expected table ids: {:?}, actual table infos' table id: {:?}, those table doesn't exist: {:?}",
+                tids,
+                table_infos.keys(),
+                tids.iter().filter(|id| !table_infos.contains_key(id)).collect::<Vec<_>>()
+            );
+        }
+
+        let group_by_table_name = group_by_table_id
+            .into_iter()
+            .filter_map(|(id, rows)| {
+                let table_name = table_infos.get(&id).map(|info| info.table_name());
+                let Some(table_name) = table_name else {
+                    warn!("Failed to get table infos for table id: {:?}", id);
+                    return None;
+                };
+                let table_name = [
+                    table_name.catalog_name,
+                    table_name.schema_name,
+                    table_name.table_name,
+                ];
+                Some((table_name, rows))
+            })
+            .collect::<HashMap<_, _>>();
+
+        // TODO(discord9): not use one lock for all tasks
+        for (_flow_id, task) in self.tasks.read().await.iter() {
```

Comment:
can we somehow make the handling logic parallel? I'm afraid this will become a bottleneck